### PR TITLE
add filepath support for create command

### PR DESCRIPTION
**Pull request by whatsadebugger**
I dont think you can add $HOME as a hard coded test because shell expansion usually gives you the absolute path.
 I did however test chainlink create with ~  / $HOME / $GOPATH they all create jobs.

### CODE REVIEW

File: cmd/client.go [hunk 1]
```diff
@@ -92,14 +92,24 @@ func (cli *Client) GetJobSpecs(c *clipkg.Context) error {
 func (cli *Client) CreateJobSpec(c *clipkg.Context) error {
 	cfg := cli.Config
 	if !c.Args().Present() {
-		return cli.errorOut(errors.New("Must pass in job JSON"))
+		return cli.errorOut(errors.New("Must pass in JSON/filepath"))
```
Comment (dimroc) [inline cmd/client.go]: Change this to `JSON or filepath`

File: cmd/client.go [hunk 2]
```diff
@@ -92,14 +92,24 @@ func (cli *Client) GetJobSpecs(c *clipkg.Context) error {
...
+	}
+	var buf *bytes.Buffer
+	if c.Args().First()[0] != '{' {
+		file, err := ioutil.ReadFile(c.Args().First())
```
Comment (dimroc) [inline cmd/client.go]: Let's have support for `~` paths. This is done with `homedir` as show here:

```
	dir, err := homedir.Expand(config.RootDir)
```

https://github.com/smartcontractkit/chainlink/blob/master/store/config.go#L41

File: cmd/client.go [hunk 3]
```diff
@@ -92,14 +92,24 @@ func (cli *Client) GetJobSpecs(c *clipkg.Context) error {
 func (cli *Client) CreateJobSpec(c *clipkg.Context) error {
 	cfg := cli.Config
 	if !c.Args().Present() {
-		return cli.errorOut(errors.New("Must pass in job JSON"))
+		return cli.errorOut(errors.New("Must pass in JSON/filepath"))
+	}
+	var buf *bytes.Buffer
+	if c.Args().First()[0] != '{' {
```
Comment (dimroc) [inline cmd/client.go]: Let's have a more robust json check. We already have `gjson` so you can do the following:
https://github.com/tidwall/gjson#validate-json

```
if !gjson.Valid(json) {
	return errors.New("invalid json")
}
```

**Comment by whatsadebugger** [comment]
Thanks for the feedback I'll add those changes later today.

**Comment by se3000** [comment]
Could you rebase to resolve the merge conflicts? It looks like CI never even started your build, and I think it's waiting on the merge conflicts to be resolved.

File: cmd/client.go [hunk 4]
```diff
@@ -92,14 +95,29 @@ func (cli *Client) GetJobSpecs(c *clipkg.Context) error {
...
+	} else {
+		dir, err := homedir.Expand(arg)
+		if err != nil {
+			return cli.errorOut(err)
+		}
+		file, err := ioutil.ReadFile(dir)
+		if err != nil {
+			return cli.errorOut(err)
+		}
+		buf = bytes.NewBuffer(file)
```
Comment (dimroc) [inline cmd/client.go]: Let's place this entire `else` into a a `fromFile` method to improve readability.

File: cmd/client_test.go [hunk 5]
```diff
@@ -89,14 +94,18 @@ func TestClient_CreateJobSpec(t *testing.T) {
...
 		errored bool
 	}{
-		{"bad input", 0, true},
+		{"{bad json}", 0, true},
+		{"bad/filepath/", 0, true},
 		{`{"initiators":[{"type":"web"}],"tasks":[{"type":"NoOp"}]}`, 1, false},
 		{`{"initiators":[{"type":"runAt","time":"2018-01-08T18:12:01.103Z"}],"tasks":[{"type":"NoOp"}]}`, 2, false},
+		{"../internal/fixtures/web/end_at_job.json", 3, false},
+		{expandGoPath("internal/fixtures/web/hello_world_job.json"), 4, false},
+		{expandGoPath("internal/fixtures/web/invalid_cron.json"), 4, true},
```
Comment (dimroc) [inline cmd/client_test.go]: I understand we might want to test the `~/some_job.json` but for now lets skip that scenario. We already have coverage and that edge case is handled by `homedir`, no need to test a third party package.

File: cmd/client_test.go [hunk 6]
```diff
@@ -89,14 +94,18 @@ func TestClient_CreateJobSpec(t *testing.T) {
...
 	}
 
 	for _, tt := range tests {
 		test := tt
-
+		fmt.Println(test.input)
```
Comment (dimroc) [inline cmd/client_test.go]: Don't forget to nuke this when you're out of WIP

File: cmd/client.go [hunk 7]
```diff
@@ -92,14 +95,24 @@ func (cli *Client) GetJobSpecs(c *clipkg.Context) error {
...
+		return cli.errorOut(errors.New("Must pass in JSON or filepath"))
+	}
+	arg := c.Args().First()
+	var buf *bytes.Buffer
+	var err error
+	if gjson.Valid(arg) {
+		buf = bytes.NewBufferString(arg)
+	} else {
+		if buf, err = fromFile(arg, cli); err != nil {
+			return cli.errorOut(err)
```
Comment (dimroc) [inline cmd/client.go]: Let's collapse

```
} else {
  if {
  }
}
```

into one else if
```
} else if {
}
```

File: cmd/client.go [hunk 8]
```diff
@@ -110,6 +123,18 @@ func (cli *Client) CreateJobSpec(c *clipkg.Context) error {
 	return cli.deserializeResponse(resp, &jobs)
 }
 
+func fromFile(arg string, cli *Client) (*bytes.Buffer, error) {
```
Comment (dimroc) [inline cmd/client.go]: `fromFile` never uses `cli *Client` so let's remove that parameter.

File: cmd/client.go [hunk 9]
```diff
@@ -92,14 +95,22 @@ func (cli *Client) GetJobSpecs(c *clipkg.Context) error {
...
-		return cli.errorOut(errors.New("Must pass in job JSON"))
+		return cli.errorOut(errors.New("Must pass in JSON or filepath"))
+	}
+	arg := c.Args().First()
+	var buf *bytes.Buffer
+	var err error
+	if gjson.Valid(arg) {
+		buf = bytes.NewBufferString(arg)
+	} else if buf, err = fromFile(arg); err != nil {
+		return cli.errorOut(err)
```
Comment (dimroc) [inline cmd/client.go]: In the event someone passes malformed json, they'll get back an error saying unable to open file. The prevent any misdirection or confusion, let's do the following for now:

```
return cli.errorOut(multierr.Append(errors.New("Must pass in valid JSON or filepath"), err))
```

We can DRY up that error in a var at the top of the file.
